File: lib/base/sheet.dart
import 'database.dart';

class TagTagsSheetDataFormatException implements Exception {
  final String errMsg;

  TagTagsSheetDataFormatException(this.errMsg);
}

class TagTagsSheetData {
  final Map _data;
  final String id;

  final String project;
  final String title;
  final int version;
  late int columns;
  late bool singlePage;
  late TagTagsGroupData identifier;
  late List<TagTagsGroupData> groups = [];

  final Map<String, TagTagsDataPoint> identComps;

  TagTagsSheetData(this.id, this.project, this.title, this.version, this._data, this.identComps) {
    if (!_data.containsKey('columns') || _data['columns'].runtimeType != int)
      throw TagTagsSheetDataFormatException(
          "sheet column count not set or not an integer");

    if (!_data.containsKey('identifier'))
      throw TagTagsSheetDataFormatException(
          "sheet identifier group not defined");

    if (!_data.containsKey('groups'))
      throw TagTagsSheetDataFormatException("sheet has no groups");

    if (_data.containsKey('single_page') && _data['single_page'].runtimeType == bool && _data['single_page'] == true)
      singlePage = true;
    else
      singlePage = false;

    columns = _data['columns'] ?? 1;
    identifier = TagTagsGroupData(_data['identifier'], true);
    for (var x = 0; x < _data['groups'].length; x++)
      groups.add(TagTagsGroupData(_data['groups'][x], false));
  }
}

class TagTagsGroupData {
  final Map _data;

  late String title;
  late String constructor;
  String? description;
  String? visibleIf;
  List<TagTagsFieldData> fields = [];

  TagTagsGroupData(this._data, bool ident) {
    if (!_data.containsKey('title') || _data['title'].runtimeType != String)
      throw TagTagsSheetDataFormatException(
          "group title not set or not a string");

    if (ident &&
        (!_data.containsKey('constructor') ||
            _data['constructor'].runtimeType != String))
      throw TagTagsSheetDataFormatException(
          "identifier group constructor not set or not a string");

    if (!_data.containsKey('fields'))
      throw TagTagsSheetDataFormatException("group has no fields");

    description = _data['description'];
    visibleIf = _data['visible_if'];
    title = _data['title'];
    constructor = _data['constructor'];
    for (var x = 0; x < _data['fields'].length; x++)
      fields.add(TagTagsFieldData(_data['fields'][x]));
  }
}

class TagTagsFieldData {
  final Map _data;

  late String id;
  late String title;
  String? description;
  late bool mandatory;
  late bool allowOther;
  late int type;
  List<String> alternatives = [];
  String? visibleIf;
  String? function;
  bool checkedIsDefault = false;
  bool barcode = false;
  bool rememberValues = false;

  TagTagsFieldData(this._data) {
    if (!_data.containsKey('id') || _data['id'].runtimeType != String)
      throw TagTagsSheetDataFormatException("field id not set or not a string");

    if (!_data.containsKey('title') || _data['title'].runtimeType != String)
      throw TagTagsSheetDataFormatException(
          "field title not set or not a string");

    if (!_data.containsKey('type') || _data['type'].runtimeType != int)
      throw TagTagsSheetDataFormatException(
          "field type not set or not an integer");

    id = _data['id'];
    title = _data['title'];
    description = _data['description'];
    mandatory = _data['mandatory'] ?? false;
    allowOther = _data['allow_other'] ?? false;
    checkedIsDefault = _data['default_checked'] ?? false;
    barcode = _data['barcode'] ?? false;
    type = _data['type'];
    visibleIf = _data['visible_if'];
    function = _data['function'];
    rememberValues = _data['remember_values'] ?? false;

    if (_data.containsKey('alternatives'))
      for (var x = 0; x < _data['alternatives'].length; x++)
        alternatives.add(_data['alternatives'][x].toString());
  }
}

File: tagserv/sheet.go
package main

import (
	"database/sql"
	"encoding/json"
	"fmt"
)

type tagTagsField struct {
	ID               string   `json:"id"`
	Title            string   `json:"title"`
	TypeID           uint     `json:"type"`
	Desc             *string  `json:"description,omitempty"`
	Mandatory        *bool    `json:"mandatory,omitempty"`
	RememberValues   *bool    `json:"rememeber_values,omitempty"`
	VisibleIf        *string  `json:"visible_if,omitempty"`
	Alternatives     []string `json:"alternatives,omitempty"`
	AllowOther       *bool    `json:"allow_other,omitempty"`
	CheckedIsDefault *bool    `json:"default_checked,omitempty"`
	BarcodeReader    *bool    `json:"barcode,omitempty"`
	Function         *string  `json:"function,omitempty"`
	FollowUpValues   []string `json:"followup_values,omitempty"`
}

type tagTagsGroup struct {
	Title       string         `json:"title"`
	Description *string        `json:"description,omitempty"`
	Fields      []tagTagsField `json:"fields"`
	Constructor string         `json:"constructor,omitempty"`
}

type tagTagsSheet struct {
	Columns    uint           `json:"columns"`
	SinglePage bool           `json:"single_page,omitempty"`
	Name       string         `json:"name"`
	Project    string         `json:"-"`
	Version    uint           `json:"-"`
	Identifier tagTagsGroup   `json:"identifier"`
	Groups     []tagTagsGroup `json:"groups"`
}

func sheetFromJSON(sheet []byte) (*tagTagsSheet, error) {
	var ns tagTagsSheet
	err := json.Unmarshal(sheet, &ns)
	if err != nil {
		return nil, err
	}

	return &ns, nil
}

func getSheetFromDB(db *sql.DB, id int) (*tagTagsSheet, error) {
	row := db.QueryRow("SELECT sheet FROM sheets WHERE id = ?", id)
	var sheet string
	if err := row.Scan(&sheet); err != nil {
		return nil, err
	}

	return sheetFromJSON([]byte(sheet))
}

func sheetBelongsToProject(db *sql.DB, project string, sheet int) bool {
	row := db.QueryRow("SELECT COUNT(*) FROM sheets WHERE id = ? AND project = ?", sheet, project)
	var count int
	if err := row.Scan(&count); err != nil {
		return false
	}

	fmt.Println(sheet, project, count)

	return count == 1
}

func (s *tagTagsSheet) getAllFields() []string {
	fields := make([]string, 0)

	for _, g := range s.Groups {
		for _, f := range g.Fields {
			fields = append(fields, f.ID)
		}
	}

	return fields
}
